Fix crash on exit

a1rwulf opened the pull request:
Second attempt after: https://github.com/xbmc/xbmc/pull/13596

This adds a Disposal-Method to the SinkFactory and let's sinks register a method for cleaning stuff that is for example static like those alsa monitors.

@FernetMenta @fritsch could you have another look?

FernetMenta commented on xbmc/cores/AudioEngine/AESinkFactory.h:
```diff
@@ -38,12 +38,14 @@ struct AESinkInfo
 
 typedef IAESink* (*CreateSink)(std::string &device, AEAudioFormat &desiredFormat);
 typedef void (*Enumerate)(AEDeviceInfoList &list, bool force);
+typedef void (*Dispose)();
```
maybe your first suggestion "cleanup" fits better here because we don't call it when the disk is disposed but when ae is shut down.

FernetMenta commented:
apart from the naming of the method it looks good
great!

a1rwulf committed:
[AudioEngine] add a cleanup method to sink factory registry

This adds the ability for sinks to register a cleanup function.
When ActiveAESink's Dispose() is called, it will call Cleaup()
on all registered Sinks.
Be aware that the provided cleanup method must be static.

a1rwulf committed:
[AudioEngine] add cleanup for alsa sink

This adds and registers a static cleanup method for ALSA sinks.
It is needed to stop the static device monitors that are specific to ALSA.

This fixes a crash when kodi is stopped - race condition caused by
the hctl monitor that fires a device changed event when windowing is
already down.

fritsch commented:
This solution is not good. It hides the real issue:
- The monitors in AESinkALSA being static
- The lifetime of the monitors is not related to AESinkALSA object itself (only static here as a workaround)
- It makes AESinkALSA a static factory

The code hides this fact.

What really should be done:
- Explicitely transfer the lifetime of the monitor to an upper layer. The monitors are a service that are obviously independend of the sink's lifetime. All this static hacks were introduced as a workaround.
- AESinkService?
- Now implicitely the static factory keeps track of these static objects running somewhere by interacting with a registered method, which then destroys some static objects somewhere no one else can access anymore.
- We don't want to clutter the sinks itself with more static members or push that workaround into the factory, do we?

bacon-cheeseburger commented:
Given fritsch's last comment, I was wondering where things stand now. Is there going to be another version of the fix that considers his points? Is the current version of the fix going to stick for now? As someone affected by this bug, I'm curious what path we're on. Thanks!

FernetMenta commented:
This version is good. It fixes the issue and keeps behaviour correct. If anybody wants to improve things, feel free to submit a pr

koying commented:
Is initializing `reg.second.cleanupFunc` left to the sinks?
I'm afraid we now crash on exit on all platforms besides ALSA because it's junk ;)

FernetMenta commented:
> Is initializing reg.second.cleanupFunc left to the sinks?

nope, it should be init to nullptr by default. sorry, missed that
I'll pr a fix later
